DashSync/shared/Models/CoinJoin/Utils/DSMasternodeGroup.m

```diff
@@ -356,25 +475,86 @@ - (void)peerConnected:(nonnull DSPeer *)peer {
         
         if (numPeers < _maxConnections) {
             [self triggerConnections];
         }
     }
+     
+    @synchronized (_pendingClosingMasternodes) {
+        DSPeer *masternode = NULL;
+        
+        for (DSPeer *mn in _pendingClosingMasternodes) {
+            if ([peer.location isEqualToString:mn.location]) {
+                masternode = mn;
+            }
+        }
+        
+        DSLog(@"[OBJ-C] CoinJoin: handling this mn peer death: %@ -> %@", peer.location, masternode != NULL ? masternode.location : @"not found in closing list");
+        
+        if (masternode) {
+            NSString *address = peer.location;
+            
+            if ([_pendingClosingMasternodes containsObject:masternode]) {
+                // if this is part of pendingClosingMasternodes, where we want to close the connection,
+                // we don't want to increase the backoff time
+                [[_backoffMap objectForKey:address] trackSuccess];
+            }
+            
+            [_pendingClosingMasternodes removeObject:masternode];
+            UInt256 sessionId = [_chain.chainManager.masternodeManager masternodeAtLocation:masternode.address port:masternode.port].providerRegistrationTransactionHash;
+            NSValue *sessionIdValue = [NSValue valueWithBytes:&sessionId objCType:@encode(UInt256)];
+            [_pendingSessions removeObject:sessionIdValue];
+            [_masternodeMap removeObjectForKey:sessionIdValue];
+            [_addressMap removeObjectForKey:masternode.location];
+        }
+        
+        [self checkMasternodesWithoutSessions];
+    }
 }
 
-- (void)peer:(nonnull DSPeer *)peer relayedPeers:(nonnull NSArray *)peers { 
+- (void)peer:(nonnull DSPeer *)peer relayedPeers:(nonnull NSArray *)peers {
     // TODO ?
 }
 
+- (void)addInactive:(DSPeer *)peer {
+    @synchronized (_inactives) {
+        // Deduplicate, handle differently than PeerGroup
+        if ([self.inactives containsObject:peer]) {
+            return;
+        }
+                
+        // do not connect to another the same peer twice
+        if ([self isNodeConnected:peer]) {
+            DSLog(@"[OBJ-C] CoinJoin: attempting to connect to the same masternode again: %@", peer.location);
+            return;
+        }
+        
+        DSBackoff *backoff = [[DSBackoff alloc] initInitialBackoff:DEFAULT_INITIAL_BACKOFF maxBackoff:DEFAULT_MAX_BACKOFF multiplier:GROUP_BACKOFF_MULTIPLIER];
+        [self.backoffMap setObject:backoff forKey:peer.location];
+        [self.inactives addObject:peer];
+        [self sortInactives];
+    }
+}
+
+- (void)sortInactives {
+    [_inactives sortUsingComparator:^NSComparisonResult(DSPeer *obj1, DSPeer *obj2) {
+        DSBackoff *backoff1 = [_backoffMap objectForKey:obj1.location];
+        DSBackoff *backoff2 = [_backoffMap objectForKey:obj2.location];
+        return [backoff1.retryTime compare:backoff2.retryTime];
+    }];
+}
+
```

Comment:
A lot of this class reflects [MasternodeGroup.java](https://github.com/dashpay/dashj/blob/feature-coinjoin/core/src/main/java/org/bitcoinj/coinjoin/utils/MasternodeGroup.java) and some of the DashJ's peer management logic, which might be unnecessary and in theory, can be replaced by deriving or including `DSPeerManager`. But at this point, I'm not familiar enough with both peer managers to do this effectively.